pkg/util/io/io.go

```diff
@@ -20,12 +20,114 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"path/filepath"
+	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/apimachinery/registered"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
+// ioTimeout is the timeout to use for io operations
+const ioTimeout time.Duration = 5
+
+// Lstat wraps os.Lstat in a timeout.
+func Lstat(path string) (os.FileInfo, error) {
+	return internalLstatTimeout(path, ioTimeout)
+}
+
+func internalLstatTimeout(path string, timeout time.Duration) (os.FileInfo, error) {
+	var (
+		timer      = time.NewTimer(ioTimeout)
+		resChannel = make(chan os.FileInfo)
+		errChannel = make(chan error)
+	)
+
+	defer timer.Stop()
+
+	go func() {
+		fileInfo, err := os.Lstat(path)
+		if err != nil {
+			errChannel <- err
+		} else {
+			resChannel <- fileInfo
+		}
+	}()
+
+	select {
+	case <-timer.C:
+		return nil, fmt.Errorf("Lstat on %v timed out after %v", path, timeout)
```

Comment:
What is causing it to hang?  I don't think that leaking goroutines is a
viable answer to this problem unless you can answer categorically what is
going on.

On Wed, Aug 24, 2016 at 6:04 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/util/io/io.go
> <https://github.com/kubernetes/kubernetes/pull/31379#discussion_r76166245>
> :
>
> > +	)
> > +
> > +	defer timer.Stop()
> > +
> > +	go func() {
> > +		fileInfo, err := os.Lstat(path)
> > +		if err != nil {
> > +			errChannel <- err
> > +		} else {
> > +			resChannel <- fileInfo
> > +		}
> > +	}()
> > +
> > +	select {
> > +	case <-timer.C:
> > +		return nil, fmt.Errorf("Lstat on %v timed out after %v", path, timeout)
>
> Good point... can you think of any alternatives? The lstat call just hangs
> in situations like these...
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31379/files/32202031260c40958c2f5c398984bb8a784e1583#r76166245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGFW5M60p6EgTQBkHjLbT31GePrmks5qjOobgaJpZM4Jsd8Y>
> .
>